# 4.7 Deriving New Variables

## This script covers the following topics:

### 1. Import Merge Data
### 2. Create Subset to Experiment/Create New Variables
### 3. Define price_label Function and Variable
### 4. Apply price_label Function to Subset
### 5. If Statements with the loc() Function
### 6. Incorporate loc() 'price_range_loc' in ords_prods_merge Dataframe 
### 7. If Statements with For-Loops
### 8. Exercise 4.7 Task

In [1]:
#Import Libraries
import os
import pandas as pd
import numpy as np

In [2]:
#Set Path
path = r'C:\Users\Josh Wattay\anaconda3\Instacart Basket Analysis'

### 1. Import Merge Data

In [3]:
#Import Data Sets
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
#Check output
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_time_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [5]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          orders_time_of_day  days_since_prior_order  product_id  \
32404854                  15                     1.0       43553   
32404855                  11                     3.0       43553   
32404856                   7                     NaN       42338   
32404857                   5                    15.0       42338   
32404858                  14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id     prices _merge  
32404854        64              7   3.699219   both  
32404855        64              7   3.699219   both  
32404856        50             19   6.898438   both  
32404857        50             19   6.898438   both  
32404858      -126             14  10.296875   both

In [6]:
#Shape Check
ords_prods_merge.shape

(32404859, 14)

### 2. Create Subset to Experiment/Create New Variables

In [7]:
df = ords_prods_merge[: 1000000]

In [8]:
#check output

df.tail()

order_id  user_id  order_number  orders_day_of_week  \
999995   2836489   164627            14                   0   
999996   1843600   164632             5                   1   
999997    733106   164632             9                   2   
999998   1650124   164632            17                   6   
999999   3378258   164632            22                   6   

        orders_time_of_day  days_since_prior_order  product_id  \
999995                  15                     6.0       30489   
999996                  19                     9.0       30489   
999997                  22                    10.0       30489   
999998                  17                    13.0       30489   
999999                  10                    15.0       30489   

        add_to_cart_order  reordered     product_name  aisle_id  \
999995                  2          0  Original Hummus        67   
999996                  2          0  Original Hummus        67   
999997                  3          1  Original Hummus        67   
999998                  1          1  Original Hummus        67   
999999                  3          1  Original Hummus        67   

        department_id  prices _merge  
999995             20     7.5   both  
999996             20     7.5   both  
999997             20     7.5   both  
999998             20     7.5   both  
999999             20     7.5   both

In [9]:
# Shape check

df.shape

(1000000, 14)

### 3. Define price_label Function and Variable

In [10]:
# If the value within the “prices” column within the given row is less than or equal to 5, then return the string “Low-range product.”

# Or else, if the value within the “prices” column within the given row is greater than 5 and less than or equal to 15, then return the string “Mid-range product.”

# Or else, if the value of the “prices” column within the given row is greater than 15, then return the string “High-range product.”

# Or else, return the string “Not enough data.”

In [11]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

### 4. Apply price_label Function to Subset

In [12]:
# Applying new price_label function to subset of data

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_19448\426245314.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [13]:
pd.options.mode.copy_on_write = True

In [14]:
df['price_range'] = df.apply(price_label, axis=1)

In [15]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [16]:
df['prices'].max()

14.8

### 5. If Statements with the loc() Function

In [17]:
# if = df.loc[df['prices'] > 15,
# then = 'price_range_loc'] = 'High-range product'

In [18]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
# if = df.loc[(df['prices'] <= 15) & (df['prices'] > 5),
# then = 'price_range_loc'] = 'Mid-range product'

In [20]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [21]:
# if = df.loc[df['prices'] <= 5,
# then = 'price_range_loc'] = 'Low-range product'

In [22]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [23]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

### 6. Incorporate loc() 'price_range_loc' in ords_prods_merge Dataframe 

In [24]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [25]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [26]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [27]:
#Value Count Frequency Check

ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [46]:
price_range_loc_result = ords_prods_merge['price_range_loc'].value_counts(dropna = False)

In [47]:
price_range_loc_result.to_clipboard()

### 7. If Statements with For-Loops

In [28]:
#Inspect for busiest day of week using Frequency Check value_counts() function

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [48]:
orders_day_of_week_table = ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

In [49]:
orders_day_of_week_table.to_clipboard()

In [29]:
# 0 - Saturday
# 1 - Sunday
# 2 - Monday
# 3 - Tuesday
# 4 - Wednesday
# 5 - Thursday
# 6 - Friday

#### It appears that Saturday is the busiest day for Orders, followed by Sunday and Friday. This is slightly intuitive given the majority of people likely have more time on the weekends to place orders when they are not working (Monday-Friday).

#### Wednesday is the slowest day, with roughly 61% of the order volume that occurs on Saturday. Tuesday, Wednesday and Thursday would be a great opportunity to run targeted ads.

In [30]:
#Creating a For-Loop to make a 'busiest_day' column

result = []

In [31]:
for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [32]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [52]:
#Combine Result with ords_prods_merge data for effective use in analysis

ords_prods_merge['busiest_day'] = result

In [53]:
# Frequency Check

ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [54]:
orders_busiest_day = ords_prods_merge['busiest_day'].value_counts(dropna = False)

In [55]:
orders_busiest_day.to_clipboard()

### 8. Exercise 4.7 Task

#### Step 1 previously completed above - see Markdown Sections 4 through 7

#### Step 2
##### Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [35]:
#Drop busiest_day column

ords_prods_merge.drop('busiest_day', axis=1, inplace=True)

In [36]:
# Verify dropped column with info check

In [37]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 15 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int32   
 1   user_id                 int32   
 2   order_number            int8    
 3   orders_day_of_week      int8    
 4   orders_time_of_day      int8    
 5   days_since_prior_order  float16 
 6   product_id              int32   
 7   add_to_cart_order       int32   
 8   reordered               int32   
 9   product_name            object  
 10  aisle_id                int8    
 11  department_id           int8    
 12  prices                  float16 
 13  _merge                  category
 14  price_range_loc         object  
dtypes: category(1), float16(2), int32(5), int8(5), object(2)
memory usage: 1.4+ GB


In [38]:
# Add shell Result prior to defining for loop

result = []

In [39]:
for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value == 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

##### Since the stakeholders want to incorporate the two busiest and slowest days, I've included day 1 (Sunday) to the "Busiest days" section, and day 3 (Tuesday) to the relabelled "Slowest days" section.

In [40]:
#Create busiest_days Column using above For-Loop result

ords_prods_merge['busiest_days'] = result

In [58]:
result

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

#### Step 3
##### Check the values of this new column for accuracy. Note any observations in markdown format.

In [41]:
#Frequency Check

ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [56]:
ords_busiest_days = ords_prods_merge['busiest_days'].value_counts(dropna = False)

In [57]:
ords_busiest_days.to_clipboard()

##### This distribution is almost a normal distribution with the majority in the regularly busy days (2 - Monday, 5- Thursday, 6- Friday).
##### There is a slight rightward skew in the direction of busiest days (0- Saturday and 1- Sunday) which is to be expected, because they are busier.

In [42]:
#Aggregate Check
12916111 + 11864412 + 7624336

32404859

In [43]:
#Shape Check to verify Aggregate Check

ords_prods_merge.shape

(32404859, 16)

##### The 32,404,859 records by adding the records in the different categories of the busiest_days column MATCH the total rows/records in our shape check of the merge dataframe. This verifies that all records are still accounted for.

In [44]:
#Check output

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_time_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

     busiest_days  
0  Regularly busy  
1    Slowest days  
2    Slowest days  
3    Slowest days  
4    Slowest days

In [59]:
ords_prods_merge.drop('price_range_loc', axis=1, inplace=True)

In [60]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          orders_time_of_day  days_since_prior_order  product_id  \
32404854                  15                     1.0       43553   
32404855                  11                     3.0       43553   
32404856                   7                     NaN       42338   
32404857                   5                    15.0       42338   
32404858                  14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id     prices _merge    busiest_days  \
32404854        64              7   3.699219   both  Regularly busy   
32404855        64              7   3.699219   both  Regularly busy   
32404856        50             19   6.898438   both  Regularly busy   
32404857        50             19   6.898438   both    Slowest days   
32404858      -126             14  10.296875   both    Slowest days   

             busiest_day  
32404854  Regularly busy  
32404855  Regularly busy  
32404856  Regularly busy  
32404857    Slowest days  
32404858    Slowest days

#### Step 4
##### When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

##### Lets start by getting value counts of the orders_time_of_day column to identify busiest and slowest hours first.

In [61]:
#Frequency check of orders_time_of_day

ords_prods_merge['orders_time_of_day'].value_counts(dropna = False)

orders_time_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

##### Given that the stakeholders want 3 groups and we have 24 hours, it makes sense that each group should have 8 individual hours assigned.
##### "Most Orders" will be: 10, 11, 14, 15, 13, 12, 16, 9
##### "Average Orders" will be: 17, 8, 18, 19, 20, 7, 21, 22
##### "Fewest Orders" will be: 6, 0, 1, 5, 2, 4, 3

In [62]:
#Create shell Result prior to For-Loop code
result = []

In [63]:
for value in ords_prods_merge["orders_time_of_day"]:
  if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
    result.append("Most Orders")
  elif value == 17 or value == 8 or value == 18 or value == 19 or value == 20 or value == 7 or value == 21 or value == 22:
    result.append("Average Orders")
  else:
    result.append("Fewest Orders")

In [64]:
ords_prods_merge['busiest_period_of_day'] = result

#### Step 5
##### Print the frequency for this new column.

In [65]:
#Frequency Check

ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: count, dtype: int64

In [66]:
ords_busiest_period = ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

ords_busiest_period.to_clipboard()

In [ ]:
#Aggregate check
21118071 + 9997651 + 1289137

In [ ]:
#Shape check to verify Aggregate check
ords_prods_merge.shape

##### Our shape rows/records of 32,404,859 MATCHES our Aggregate check of 32,404,859 accounting for the values discovered in the Frequency check.

In [ ]:
#Check output
ords_prods_merge.head()

In [ ]:
ords_prods_merge.tail()

#### Step 6
##### Ensured notebook is clean and structured and that code is well commented.

#### Step 7
##### Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4.7.pkl'))